include/lsst/ip/diffim/DipoleAlgorithms.h

```diff
@@ -11,232 +11,375 @@
+        std::string const & doc);
+
+    // A typedef for the FunctorKey which returns the result of this algorithm
+    typedef lsst::meas::base::FluxResultKey ResultKey;
     /// @brief Return the standard flux keys registered by this algorithm.
-    KeyTuple const & getPositiveKeys() const { return _positiveKeys; }
-    KeyTuple const & getNegativeKeys() const { return _negativeKeys; }
-
-protected:
+    ResultKey const & getPositiveKeys() const { return _positiveKeys; }
+    ResultKey const & getNegativeKeys() const { return _negativeKeys; }
 
-    /// @brief Initialize with a manually-constructed key tuple.
-    DipoleFluxAlgorithm(DipoleFluxControl const & ctrl, KeyTuple const & positiveKeys,
-                        KeyTuple const & negativeKeys);
+    virtual void measure(
+        afw::table::SourceRecord & measRecord,
+        afw::image::Exposure<float> const & exposure
+    ) const = 0;
 
-    /// @brief Initialize using afw::table::addFlux field to fill out repetitive descriptions.
-    DipoleFluxAlgorithm(DipoleFluxControl const & ctrl, 
-                        afw::table::Schema & schema, char const * doc);
+    virtual void fail(
+        afw::table::SourceRecord & measRecord,
+        lsst::meas::base::MeasurementError * error=NULL
+    ) const = 0;
 
-private:
-    KeyTuple _positiveKeys;
-    KeyTuple _negativeKeys;
-};
+protected:
 
-class DipoleFluxControl : public meas::algorithms::AlgorithmControl {
-public:
+    /// @brief Initialize with a manually-constructed result key.
+    DipoleFluxAlgorithm(Control const & ctrl, std::string const & name,
+                        afw::table::Schema & schema, std::string const & doc,
+                        ResultKey const & positiveKeys, ResultKey const & negativeKeys);
 
-    virtual PTR(DipoleFluxControl) clone() const { 
-        return boost::static_pointer_cast<DipoleFluxControl>(_clone()); }
+    Control _ctrl;
+    lsst::meas::base::FluxResultKey _fluxResultKey;
+    lsst::meas::base::FlagHandler _flagHandler;
+    lsst::meas::base::SafeCentroidExtractor _centroidExtractor;
 
-    virtual PTR(DipoleFluxAlgorithm) makeAlgorithm(
-        afw::table::Schema & schema,
-        PTR(daf::base::PropertyList) const & metadata = PTR(daf::base::PropertyList)(),
-        meas::algorithms::AlgorithmMap const & others = meas::algorithms::AlgorithmMap()
-    ) const {
-        return boost::static_pointer_cast<DipoleFluxAlgorithm>(
-            _makeAlgorithm(schema, metadata, others));
-    }
-
-protected:
-    explicit DipoleFluxControl(std::string const & name_, double priority=2.0) :
-        AlgorithmControl(name_, priority) {}
+    ResultKey _positiveKeys;
+    ResultKey _negativeKeys;
 };
 
 inline DipoleCentroidAlgorithm::DipoleCentroidAlgorithm(
-    DipoleCentroidControl const & ctrl, KeyTuple const & positiveKeys, KeyTuple const & negativeKeys
+    Control const & ctrl, std::string const & name, afw::table::Schema & schema, std::string const & doc
```

Comment:
Why not use DipoleCentroidControl?  You do it below with DipoleFluxControl.  It seems more readable to me.